crates/bevy_ecs/src/schedule_v3/executor/single_threaded.rs

```diff
@@ -0,0 +1,161 @@
+#[cfg(feature = "trace")]
+use bevy_utils::tracing::info_span;
+use fixedbitset::FixedBitSet;
+
+use crate::{
+    schedule_v3::{
+        is_apply_system_buffers, ticks_since, ExecutorKind, SystemExecutor, SystemSchedule,
+    },
+    world::World,
+};
+
+/// Runs the schedule using a single thread.
+#[derive(Default)]
+pub struct SingleThreadedExecutor {
+    /// System sets whose conditions have either been evaluated or skipped.
+    completed_sets: FixedBitSet,
+    /// Systems that have run or been skipped.
+    completed_systems: FixedBitSet,
+    /// Systems that have run but have not had their buffers applied.
+    unapplied_systems: FixedBitSet,
+}
+
+impl SystemExecutor for SingleThreadedExecutor {
+    fn kind(&self) -> ExecutorKind {
+        ExecutorKind::SingleThreaded
+    }
+
+    fn init(&mut self, schedule: &SystemSchedule) {
+        // pre-allocate space
+        let sys_count = schedule.system_ids.len();
+        let set_count = schedule.set_ids.len();
+        self.completed_sets = FixedBitSet::with_capacity(set_count);
+        self.completed_systems = FixedBitSet::with_capacity(sys_count);
+        self.unapplied_systems = FixedBitSet::with_capacity(sys_count);
+    }
+
+    fn run(&mut self, schedule: &mut SystemSchedule, world: &mut World) {
+        for sys_idx in 0..schedule.systems.len() {
+            if self.completed_systems.contains(sys_idx) {
+                continue;
+            }
+
+            #[cfg(feature = "trace")]
+            let name = schedule.systems[sys_idx].get_mut().name();
+            #[cfg(feature = "trace")]
+            let should_run_span = info_span!("check_conditions", name = &*name).entered();
+
+            // evaluate conditions
+            let mut should_run = true;
+            let world_tick = world.change_tick();
+
+            // evaluate system set conditions in hierarchical order
+            for set_idx in schedule.sets_of_systems[sys_idx].ones() {
+                if self.completed_sets.contains(set_idx) {
+                    continue;
+                }
+
+                let set_conditions = schedule.set_conditions[set_idx].get_mut();
+
+                // if any condition fails, we need to restore their change ticks
+                let saved_tick = set_conditions
```

Comment:
(Disclaimer: I'd also suggest reading my response to your comment below first. My interpretation of how "change-detecting conditions" were supposed to work was "return `true` if <thing> changed since last time the guarded system (set) ran".)

So atm, it's actually "youngest" now (just `min` was incorrect). If we keep with my interpretation, I'm pretty sure it has to be either "youngest" (in which case, yeah we could just take first tick) or we'd restore each condition's tick exactly (`O(n)`).

One "issue" is you can call `configure_set` multiple times, so you can add conditions to a system set at any time. I decided that a newly-added "change-detecting condition" should respect "since last time the system set ran" and have its change tick set accordingly[^1].


[^1]: I guess there's a problem atm if you're giving the set its first condition (no tick to inherit).